File: SubArraySum.cs
//Time Complexity: O(n)
//Space Complexity: O(n)
public class Solution {
    public int SubarraySum(int[] nums, int k) {
        Dictionary<int, int> prefixSumCount = new Dictionary<int, int>();
        prefixSumCount[0] = 1; // Base case: to handle exact matches with k

        int prefixSum = 0;
        int count = 0;

        foreach (int num in nums) {
            prefixSum += num; // Update running sum

            if (prefixSumCount.ContainsKey(prefixSum - k)) {
                count += prefixSumCount[prefixSum - k]; // Add count of subarrays
            }

            if (prefixSumCount.ContainsKey(prefixSum)) {
                prefixSumCount[prefixSum]++;
            } else {
                prefixSumCount[prefixSum] = 1;
            }
        }

        return count;
    }
}

File: subarraysum.cpp
// Time Complexity : O(n)
// Space Complexity : O(n) in worst case stroing all prefix sums in hashmap
// Did this code successfully run on Leetcode : Yes

// //Approach:
// Calculate runningSum at each index. 
// Check if hashmap has the compliment already. If it does  add to that value. 
// Increment occurance

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int rSum = 0; 
        unordered_map<int, int> map;
        int count = 0; 
        map[0] = 1;

        for(int i = 0; i < nums.size(); i++){
            rSum += nums[i];
            if(map.find(rSum - k) != map.end()){
                count += map[rSum - k];
            }
            map[rSum]++;
        }
        return count;
    }
};